### PR TITLE
ssc barebones

### PR DESCRIPTION
**Pull request by eithompson**
A barebones Sylloge implementation on the SSC corpus. Uses case-insensitive search to find relevant files (and the relevant paragraphs within them). A selection is a paragraph containing a match to the search term, in addition to its adjacent paragraphs. If two selections in the same source document are adjacent, they are merged as part of the same selection.

Example: say one letter represents a paragraph, with `m = match`, `n = no match`, and `a = adjacent` (to a match). The bracketed paragraphs are grouped together as selections.
```
n n n [a m a] n n n n [a m a a m a] n [a m a] n n 
```

**See `example_output.md`** for the results when run with `example_call.txt` ("Roman Empire" search term).

### CODE REVIEW

File: barebones/sylloge.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        md_out = md_out + bolded_sels + "\n"
+    return md_out
+
+# build markdown.
+md = "# " + search_term + "\n\n"
+for filename in listdir(corpus_dir):
+    parsed = parse_file(corpus_dir + "/" + filename, corpus)
+    selections = single_source_selections(parsed["body"], search_term) 
+    if selections is not None:
+        url = parsed["url"][0:-1] # don't want newline
```
Comment (eindiran) [inline barebones/sylloge.py:96]: Consider using `parsed["url"].strip()` or `parsed["url"].rstrip()` instead, unless you're sure that there is always a trailing newline and you want to keep other trailing whitespace.

File: barebones/sylloge.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            selections_list.append([paragraph])
+
+    return selections_list
+
+def bold_matches(text, search_term):
+    '''
+    We've already searched through the text to find our selections in the first place
+    so it's a little inefficient to do it again. Fast enough though.
+    Have to search case-insensitive but return original case.
+    This implementation sucks - would rather regex: case-insensitive and capture group?
```
Comment (eindiran) [inline barebones/sylloge.py:63]: I think your comment is probably correct re this being easier/faster with a regex.

File: barebones/sylloge.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    return parsed_dict
+
+def single_source_selections(content, search_term):
+    '''
+    Takes a list of one source's paragraphs.
+    Outputs a list of lists of the selections that match the search term, context included.
+    Adjacent hit-paragraphs are concatenated into one selection.
+    '''
+    hits = [idx for idx, val in enumerate(content) if search_term.lower() in val.lower()]
+    if hits == []:
```
Comment (eindiran) [inline barebones/sylloge.py:37]: Since empty list is falsey, you can do `if not hits:`

File: barebones/ssc.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+def ssc_parse(filepath):
+    '''
+    Returns dictionary containing title, url, and body.
+    Title and url are strings. Body is a list of strings (paragraphs).
+    '''
+
+    with open(filepath) as f:
```
Comment (eindiran) [inline barebones/ssc.py:7]: I recommend specifying the encoding here, just as a sanity measure:
`with open(filepath, encoding='utf-8') as f:`

File: barebones/sylloge.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+md = "# " + search_term + "\n\n"
+for filename in listdir(corpus_dir):
+    parsed = parse_file(corpus_dir + "/" + filename, corpus)
+    selections = single_source_selections(parsed["body"], search_term) 
+    if selections is not None:
+        url = parsed["url"][0:-1] # don't want newline
+        title = parsed["title"][0:-1] # don't want newline
+        md = md + "## " + "[" + title + "](" + url + ")\n\n" + build_source_md(selections)
+
+with open(output_filepath, "w") as f:
```
Comment (eindiran) [inline barebones/sylloge.py:100]: Consider adding the encoding here as well.

**Comment by eindiran** [comment]
I added some comments, but most of it is stylistic or personal things. The code looks very good.

File: barebones/sylloge.py (hunk 3 above)
Comment (eithompson) [reply barebones/sylloge.py:37]: Why though? Is it so pythonic that people would prefer it to the equality check? Seems way less clear.

Comment (eithompson) [reply barebones/sylloge.py:37]: @eindiran

File: barebones/ssc.py (hunk 4 above)
Comment (eithompson) [reply barebones/ssc.py]: @eindiran also: does this specify "here is what the encoding of the file is" or "here is what I want you to use as encoding for this file"

Comment (eithompson) [reply barebones/ssc.py]: OK, [looks like](https://docs.python.org/3/library/functions.html#open) it's the latter.

File: barebones/sylloge.py (hunk 3 above)
Comment (eindiran) [reply barebones/sylloge.py]: Sorry, haven't been on in the last few days so I didn't see this. I agree it is less clear that you are checking for an empty list, but it is generally seen as the canonical way to do it. The main exception is if you want to do a comparison that excludes other falsey values (`None`, `False`, empty string), but is true for empty list. Using the "implicit booleaness" of a list is considered Pythonic, despite the loss of explicitness. There is some debate in the comments about this here for example: https://stackoverflow.com/questions/53513/how-do-i-check-if-a-list-is-empty

**Pull request merged**